File: ui/models/course.ts
import { Topic } from "./topic";

export type Category = {
  id: string;
  name: string;
};

export type Course = {
  id: string;
  title: string;
  description: string | null;
  imageUrl: string | null;
  price: number;
  category: string;
  level: string;
  studentIds: number[];
  sections: Section[];
  isPublished: boolean;
};

export type Section = {
  id: string;
  courseId: string;
  title: string;
  description: string | null;
  position: number;
  topics: Topic[];
};

File: server/domain/course.go
package domain

import "github.com/gofrs/uuid/v5"

type Course struct {
	ID          uuid.UUID `json:"id" gorm:"type:uuid;default:uuid_generate_v4();primaryKey"`
	Title       string    `json:"title" gorm:"not null"`
	Description *string   `json:"description" gorm:"type:text"`
	ImageUrl    *string   `json:"imageURL" gorm:"type:text"`
	Price       *float64  `json:"price" gorm:"default:null"`
	Category    *string   `json:"category" gorm:"type:text"`
	Level       *string   `json:"level" gorm:"type:text"`
	IsPublished bool      `gorm:"not null;default:false"`

	Sections  []Section
	Resources []AttachedFile

	Students []User `gorm:"many2many:user_courses"`

	Creator   User      `json:"-" gorm:"foreignKey:CreatorID;constraint:OnDelete:CASCADE"`
	CreatorID uuid.UUID `json:"-" gorm:"not null;index"`
}
